{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-07T12:52:49.879995Z",
     "start_time": "2024-08-07T12:52:49.218029Z"
    }
   },
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "sys.path.append('..')\n",
    "import medicc\n",
    "import medicc.bootstrap"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data loading"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-07T12:52:52.807370Z",
     "start_time": "2024-08-07T12:52:52.750976Z"
    }
   },
   "source": [
    "# By setting the numpy seed, the created bootstrap datasets will always be the same\n",
    "# Comment this line out if you want to create different datasets each time the notebook is run\n",
    "np.random.seed(42)\n",
    "\n",
    "##%%\n",
    "data_folder = os.path.join(\"../examples/output_gundem_et_al_2015\")\n",
    "\n",
    "main_tree = medicc.io.import_tree(os.path.join(data_folder, \"PTX011_final_tree.new\"), 'diploid')\n",
    "main_df = medicc.io.read_and_parse_input_data(os.path.join(data_folder, \"PTX011_final_cn_profiles.tsv\"))\n",
    "\n",
    "# remove internal nodes from df\n",
    "main_df = main_df.loc[main_df.index.get_level_values('sample_id').map(lambda x: 'internal' not in x)]\n",
    "\n",
    "labels = {'diploid': 'Diploid'}\n",
    "for label in main_df.reset_index()['sample_id']:\n",
    "    if 'diploid' not in label and 'internal' not in label:\n",
    "        labels[label] = '_'.join([label.split('_')[1].split('-')[0], label.split('_')[-1]])\n"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-07 14:52:52,802 - WARNING medicc.io:\tTotal of 1.4e+08 bp gaps in the segmentation. Large gaps might affect the performance of MEDICC2.\n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## chromosome-wise bootstrap"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-07T12:53:07.455427Z",
     "start_time": "2024-08-07T12:53:07.339584Z"
    }
   },
   "source": [
    "N_bootstrap = 20\n",
    "\n",
    "bootstrap_trees_df, support_tree = medicc.bootstrap.run_bootstrap(main_df, main_tree, N_bootstrap=N_bootstrap, \n",
    "                                                                  method='chr-wise', normal_name='diploid',)\n",
    "\n",
    "print('{} chromosome-wise bootstrap datasets created'.format(N_bootstrap))\n",
    "print('{} individual trees'.format(len(bootstrap_trees_df)))\n",
    "print('The original tree was recreated {} times ({:.1f}%)'.format(bootstrap_trees_df.loc[bootstrap_trees_df['note'] == 'original', 'count'].iloc[0],\n",
    "                                                                  100*bootstrap_trees_df.loc[bootstrap_trees_df['note'] == 'original', 'freq'].iloc[0]))\n"
   ],
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "int() argument must be a string, a bytes-like object or a real number, not 'Tree'",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[4], line 3\u001B[0m\n\u001B[1;32m      1\u001B[0m N_bootstrap \u001B[38;5;241m=\u001B[39m \u001B[38;5;241m20\u001B[39m\n\u001B[0;32m----> 3\u001B[0m bootstrap_trees_df, support_tree \u001B[38;5;241m=\u001B[39m medicc\u001B[38;5;241m.\u001B[39mbootstrap\u001B[38;5;241m.\u001B[39mrun_bootstrap(main_df, main_tree, N_bootstrap\u001B[38;5;241m=\u001B[39mN_bootstrap, \n\u001B[1;32m      4\u001B[0m                                                                   method\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mchr-wise\u001B[39m\u001B[38;5;124m'\u001B[39m, normal_name\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mdiploid\u001B[39m\u001B[38;5;124m'\u001B[39m,)\n\u001B[1;32m      6\u001B[0m \u001B[38;5;28mprint\u001B[39m(\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m chromosome-wise bootstrap datasets created\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;241m.\u001B[39mformat(N_bootstrap))\n\u001B[1;32m      7\u001B[0m \u001B[38;5;28mprint\u001B[39m(\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m individual trees\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;241m.\u001B[39mformat(\u001B[38;5;28mlen\u001B[39m(bootstrap_trees_df)))\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/bootstrap.py:212\u001B[0m, in \u001B[0;36mrun_bootstrap\u001B[0;34m(input_df, maxcn, original_tree, N_bootstrap, method, wgd, seed, normal_name, show_progress, n_cores)\u001B[0m\n\u001B[1;32m    210\u001B[0m     initial_trees \u001B[38;5;241m=\u001B[39m []\n\u001B[1;32m    211\u001B[0m     \u001B[38;5;28;01mfor\u001B[39;00m i \u001B[38;5;129;01min\u001B[39;00m tqdm(\u001B[38;5;28mrange\u001B[39m(N_bootstrap), disable\u001B[38;5;241m=\u001B[39m\u001B[38;5;129;01mnot\u001B[39;00m show_progress):\n\u001B[0;32m--> 212\u001B[0m         cur_tree \u001B[38;5;241m=\u001B[39m _single_bootstrap_run(\n\u001B[1;32m    213\u001B[0m             input_df, fst, maxcn, bootstrap_method, i, N_bootstrap, normal_name)\n\u001B[1;32m    214\u001B[0m         initial_trees\u001B[38;5;241m.\u001B[39mappend(cur_tree)\n\u001B[1;32m    216\u001B[0m \u001B[38;5;66;03m# delete duplicate trees\u001B[39;00m\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/bootstrap.py:133\u001B[0m, in \u001B[0;36m_single_bootstrap_run\u001B[0;34m(input_df, fst, maxcn, bootstrap_method, i, N_bootstrap, normal_name)\u001B[0m\n\u001B[1;32m    131\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21m_single_bootstrap_run\u001B[39m(input_df, fst, maxcn, bootstrap_method, i, N_bootstrap, normal_name\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mdiploid\u001B[39m\u001B[38;5;124m'\u001B[39m):\n\u001B[1;32m    132\u001B[0m     cur_df \u001B[38;5;241m=\u001B[39m bootstrap_method(input_df)\n\u001B[0;32m--> 133\u001B[0m     _, _, _, cur_final_tree, _, _ \u001B[38;5;241m=\u001B[39m main(\n\u001B[1;32m    134\u001B[0m         cur_df,\n\u001B[1;32m    135\u001B[0m         fst,\n\u001B[1;32m    136\u001B[0m         maxcn,\n\u001B[1;32m    137\u001B[0m         normal_name,\n\u001B[1;32m    138\u001B[0m         input_tree\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[1;32m    139\u001B[0m         ancestral_reconstruction\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mFalse\u001B[39;00m,\n\u001B[1;32m    140\u001B[0m         chr_separator\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mX\u001B[39m\u001B[38;5;124m'\u001B[39m)\n\u001B[1;32m    142\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m (i\u001B[38;5;241m+\u001B[39m\u001B[38;5;241m1\u001B[39m) \u001B[38;5;241m%\u001B[39m (N_bootstrap\u001B[38;5;241m/\u001B[39m\u001B[38;5;241m/\u001B[39m\u001B[38;5;241m5\u001B[39m) \u001B[38;5;241m==\u001B[39m \u001B[38;5;241m0\u001B[39m:\n\u001B[1;32m    143\u001B[0m         logger\u001B[38;5;241m.\u001B[39minfo(\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m/\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m (\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m%\u001B[39m\u001B[38;5;124m) bootstrap runs completed\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;241m.\u001B[39mformat(i \u001B[38;5;241m+\u001B[39m \u001B[38;5;241m1\u001B[39m, N_bootstrap, \n\u001B[1;32m    144\u001B[0m                                                                   \u001B[38;5;28mint\u001B[39m((i\u001B[38;5;241m+\u001B[39m\u001B[38;5;241m1\u001B[39m)\u001B[38;5;241m/\u001B[39mN_bootstrap\u001B[38;5;241m*\u001B[39m\u001B[38;5;241m100\u001B[39m)))\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/core.py:59\u001B[0m, in \u001B[0;36mmain\u001B[0;34m(input_df, asymm_fst, maxcn, normal_name, input_tree, ancestral_reconstruction, chr_separator, prune_weight, allele_columns, wgd_x2, no_wgd, total_cn, n_cores, reconstruct_events)\u001B[0m\n\u001B[1;32m     54\u001B[0m     pairwise_distances \u001B[38;5;241m=\u001B[39m parallelization_calc_pairwise_distance_shrink_idea(sample_labels, asymm_fst, FSA_dict,\n\u001B[1;32m     55\u001B[0m                                                                             symbol_table, maxcn, chr_separator,\n\u001B[1;32m     56\u001B[0m                                                                             n_cores)\n\u001B[1;32m     57\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m     58\u001B[0m     \u001B[38;5;66;03m# pairwise_distances = calc_pairwise_distance_matrix(asymm_fst, FSA_dict)\u001B[39;00m\n\u001B[0;32m---> 59\u001B[0m     pairwise_distances \u001B[38;5;241m=\u001B[39m calc_pairwise_distance_shrink_idea(asymm_fst, FSA_dict, symbol_table, maxcn,\n\u001B[1;32m     60\u001B[0m                                                             chr_separator)\n\u001B[1;32m     62\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m (pairwise_distances \u001B[38;5;241m==\u001B[39m np\u001B[38;5;241m.\u001B[39minf)\u001B[38;5;241m.\u001B[39many()\u001B[38;5;241m.\u001B[39many():\n\u001B[1;32m     63\u001B[0m     affected_pairs \u001B[38;5;241m=\u001B[39m [(pairwise_distances\u001B[38;5;241m.\u001B[39mindex[s1], pairwise_distances\u001B[38;5;241m.\u001B[39mindex[s2])\n\u001B[1;32m     64\u001B[0m                       \u001B[38;5;28;01mfor\u001B[39;00m s1, s2 \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mzip\u001B[39m(\u001B[38;5;241m*\u001B[39mnp\u001B[38;5;241m.\u001B[39mwhere((pairwise_distances \u001B[38;5;241m==\u001B[39m np\u001B[38;5;241m.\u001B[39minf)))]\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/core.py:412\u001B[0m, in \u001B[0;36mcalc_pairwise_distance_shrink_idea\u001B[0;34m(model_fst, fsa_dict, symbol_table, max_chr_num, chrom_sep, parallel_run)\u001B[0m\n\u001B[1;32m    409\u001B[0m sample_a_fst \u001B[38;5;241m=\u001B[39m fsa_dict[sample_a]\n\u001B[1;32m    410\u001B[0m sample_b_fst \u001B[38;5;241m=\u001B[39m fsa_dict[sample_b]\n\u001B[0;32m--> 412\u001B[0m sample_a_shrink_fst, sample_b_shrink_fst \u001B[38;5;241m=\u001B[39m find_common_copy_number_profile_fsa(sample_a_fst, sample_b_fst, symbol_table, max_chr_num, chrom_sep)\n\u001B[1;32m    414\u001B[0m cur_dist \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mfloat\u001B[39m(fstlib\u001B[38;5;241m.\u001B[39mkernel_score(model_fst, sample_a_shrink_fst, sample_b_shrink_fst))\n\u001B[1;32m    415\u001B[0m pdm\u001B[38;5;241m.\u001B[39mloc[sample_a, sample_b] \u001B[38;5;241m=\u001B[39m cur_dist\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/core.py:328\u001B[0m, in \u001B[0;36mfind_common_copy_number_profile_fsa\u001B[0;34m(profile_1_fsa, profile_2_fsa, symbol_table, max_chr_num, chrom_sep)\u001B[0m\n\u001B[1;32m    325\u001B[0m profile_1_shrink \u001B[38;5;241m=\u001B[39m []\n\u001B[1;32m    326\u001B[0m profile_2_shrink \u001B[38;5;241m=\u001B[39m []\n\u001B[0;32m--> 328\u001B[0m max_chr_num \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mint\u001B[39m(max_chr_num)\n\u001B[1;32m    330\u001B[0m prev_label_1 \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m\n\u001B[1;32m    331\u001B[0m prev_label_2 \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m\n",
      "\u001B[0;31mTypeError\u001B[0m: int() argument must be a string, a bytes-like object or a real number, not 'Tree'"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-07T12:53:10.143853Z",
     "start_time": "2024-08-07T12:53:10.102397Z"
    }
   },
   "source": [
    "##%% Show the most prominent trees\n",
    "fig, axs = plt.subplots(figsize=(21, 7), ncols=3)\n",
    "for ax, (i, row) in zip(axs, bootstrap_trees_df.iterrows()):\n",
    "    medicc.plot.plot_tree(row['tree'], \n",
    "                          label_func=lambda label: labels.get(label, label),\n",
    "                          ax=ax,\n",
    "                          title='{} ({:.0f}%)'.format(row['note'], 100*row['freq']))\n",
    "\n",
    "fig.suptitle('Most common bootstrap trees', y=1.0, fontsize=25, weight='bold')\n",
    "plt.show()\n"
   ],
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'bootstrap_trees_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mNameError\u001B[0m                                 Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[5], line 3\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;66;03m##%% Show the most prominent trees\u001B[39;00m\n\u001B[1;32m      2\u001B[0m fig, axs \u001B[38;5;241m=\u001B[39m plt\u001B[38;5;241m.\u001B[39msubplots(figsize\u001B[38;5;241m=\u001B[39m(\u001B[38;5;241m21\u001B[39m, \u001B[38;5;241m7\u001B[39m), ncols\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m3\u001B[39m)\n\u001B[0;32m----> 3\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m ax, (i, row) \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mzip\u001B[39m(axs, bootstrap_trees_df\u001B[38;5;241m.\u001B[39miterrows()):\n\u001B[1;32m      4\u001B[0m     medicc\u001B[38;5;241m.\u001B[39mplot\u001B[38;5;241m.\u001B[39mplot_tree(row[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mtree\u001B[39m\u001B[38;5;124m'\u001B[39m], \n\u001B[1;32m      5\u001B[0m                           label_func\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mlambda\u001B[39;00m label: labels\u001B[38;5;241m.\u001B[39mget(label, label),\n\u001B[1;32m      6\u001B[0m                           ax\u001B[38;5;241m=\u001B[39max,\n\u001B[1;32m      7\u001B[0m                           title\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m (\u001B[39m\u001B[38;5;132;01m{:.0f}\u001B[39;00m\u001B[38;5;124m%\u001B[39m\u001B[38;5;124m)\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;241m.\u001B[39mformat(row[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mnote\u001B[39m\u001B[38;5;124m'\u001B[39m], \u001B[38;5;241m100\u001B[39m\u001B[38;5;241m*\u001B[39mrow[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mfreq\u001B[39m\u001B[38;5;124m'\u001B[39m]))\n\u001B[1;32m      9\u001B[0m fig\u001B[38;5;241m.\u001B[39msuptitle(\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mMost common bootstrap trees\u001B[39m\u001B[38;5;124m'\u001B[39m, y\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m1.0\u001B[39m, fontsize\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m25\u001B[39m, weight\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mbold\u001B[39m\u001B[38;5;124m'\u001B[39m)\n",
      "\u001B[0;31mNameError\u001B[0m: name 'bootstrap_trees_df' is not defined"
     ]
    }
   ],
   "execution_count": 5
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-07T12:53:11.249385Z",
     "start_time": "2024-08-07T12:53:11.219396Z"
    }
   },
   "source": [
    "##%% Show the support for the individual branches of the main tree\n",
    "fig, ax = plt.subplots(figsize=(12, 12))\n",
    "fig = medicc.plot.plot_tree(support_tree,\n",
    "                            title='support tree',\n",
    "                            label_func=lambda label: labels.get(label, label),\n",
    "                            show_branch_lengths=True,\n",
    "                            show_branch_support=True,\n",
    "                            ax=ax)\n",
    "plt.show()\n",
    "\n"
   ],
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'support_tree' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mNameError\u001B[0m                                 Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[6], line 3\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;66;03m##%% Show the support for the individual branches of the main tree\u001B[39;00m\n\u001B[1;32m      2\u001B[0m fig, ax \u001B[38;5;241m=\u001B[39m plt\u001B[38;5;241m.\u001B[39msubplots(figsize\u001B[38;5;241m=\u001B[39m(\u001B[38;5;241m12\u001B[39m, \u001B[38;5;241m12\u001B[39m))\n\u001B[0;32m----> 3\u001B[0m fig \u001B[38;5;241m=\u001B[39m medicc\u001B[38;5;241m.\u001B[39mplot\u001B[38;5;241m.\u001B[39mplot_tree(support_tree,\n\u001B[1;32m      4\u001B[0m                             title\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124msupport tree\u001B[39m\u001B[38;5;124m'\u001B[39m,\n\u001B[1;32m      5\u001B[0m                             label_func\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mlambda\u001B[39;00m label: labels\u001B[38;5;241m.\u001B[39mget(label, label),\n\u001B[1;32m      6\u001B[0m                             show_branch_lengths\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mTrue\u001B[39;00m,\n\u001B[1;32m      7\u001B[0m                             show_branch_support\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mTrue\u001B[39;00m,\n\u001B[1;32m      8\u001B[0m                             ax\u001B[38;5;241m=\u001B[39max)\n\u001B[1;32m      9\u001B[0m plt\u001B[38;5;241m.\u001B[39mshow()\n",
      "\u001B[0;31mNameError\u001B[0m: name 'support_tree' is not defined"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## segment-wise jackknife"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-07T12:53:17.798923Z",
     "start_time": "2024-08-07T12:53:17.692130Z"
    }
   },
   "source": [
    "N_bootstrap = 20\n",
    "\n",
    "bootstrap_trees_df, support_tree = medicc.bootstrap.run_bootstrap(main_df, main_tree,\n",
    "                                                                  N_bootstrap=N_bootstrap, method='segment-wise')\n",
    "\n",
    "print('{} segment-wise jackknife datasets created'.format(N_bootstrap))\n",
    "print('{} individual trees'.format(len(bootstrap_trees_df)))\n",
    "print('The original tree was recreated {} times ({:.1f}%)'.format(bootstrap_trees_df.loc[bootstrap_trees_df['note'] == 'original', 'count'].iloc[0],\n",
    "                                                                  100*bootstrap_trees_df.loc[bootstrap_trees_df['note'] == 'original', 'freq'].iloc[0]))\n"
   ],
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "int() argument must be a string, a bytes-like object or a real number, not 'Tree'",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[7], line 3\u001B[0m\n\u001B[1;32m      1\u001B[0m N_bootstrap \u001B[38;5;241m=\u001B[39m \u001B[38;5;241m20\u001B[39m\n\u001B[0;32m----> 3\u001B[0m bootstrap_trees_df, support_tree \u001B[38;5;241m=\u001B[39m medicc\u001B[38;5;241m.\u001B[39mbootstrap\u001B[38;5;241m.\u001B[39mrun_bootstrap(main_df, main_tree,\n\u001B[1;32m      4\u001B[0m                                                                   N_bootstrap\u001B[38;5;241m=\u001B[39mN_bootstrap, method\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124msegment-wise\u001B[39m\u001B[38;5;124m'\u001B[39m)\n\u001B[1;32m      6\u001B[0m \u001B[38;5;28mprint\u001B[39m(\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m segment-wise jackknife datasets created\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;241m.\u001B[39mformat(N_bootstrap))\n\u001B[1;32m      7\u001B[0m \u001B[38;5;28mprint\u001B[39m(\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m individual trees\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;241m.\u001B[39mformat(\u001B[38;5;28mlen\u001B[39m(bootstrap_trees_df)))\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/bootstrap.py:212\u001B[0m, in \u001B[0;36mrun_bootstrap\u001B[0;34m(input_df, maxcn, original_tree, N_bootstrap, method, wgd, seed, normal_name, show_progress, n_cores)\u001B[0m\n\u001B[1;32m    210\u001B[0m     initial_trees \u001B[38;5;241m=\u001B[39m []\n\u001B[1;32m    211\u001B[0m     \u001B[38;5;28;01mfor\u001B[39;00m i \u001B[38;5;129;01min\u001B[39;00m tqdm(\u001B[38;5;28mrange\u001B[39m(N_bootstrap), disable\u001B[38;5;241m=\u001B[39m\u001B[38;5;129;01mnot\u001B[39;00m show_progress):\n\u001B[0;32m--> 212\u001B[0m         cur_tree \u001B[38;5;241m=\u001B[39m _single_bootstrap_run(\n\u001B[1;32m    213\u001B[0m             input_df, fst, maxcn, bootstrap_method, i, N_bootstrap, normal_name)\n\u001B[1;32m    214\u001B[0m         initial_trees\u001B[38;5;241m.\u001B[39mappend(cur_tree)\n\u001B[1;32m    216\u001B[0m \u001B[38;5;66;03m# delete duplicate trees\u001B[39;00m\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/bootstrap.py:133\u001B[0m, in \u001B[0;36m_single_bootstrap_run\u001B[0;34m(input_df, fst, maxcn, bootstrap_method, i, N_bootstrap, normal_name)\u001B[0m\n\u001B[1;32m    131\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21m_single_bootstrap_run\u001B[39m(input_df, fst, maxcn, bootstrap_method, i, N_bootstrap, normal_name\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mdiploid\u001B[39m\u001B[38;5;124m'\u001B[39m):\n\u001B[1;32m    132\u001B[0m     cur_df \u001B[38;5;241m=\u001B[39m bootstrap_method(input_df)\n\u001B[0;32m--> 133\u001B[0m     _, _, _, cur_final_tree, _, _ \u001B[38;5;241m=\u001B[39m main(\n\u001B[1;32m    134\u001B[0m         cur_df,\n\u001B[1;32m    135\u001B[0m         fst,\n\u001B[1;32m    136\u001B[0m         maxcn,\n\u001B[1;32m    137\u001B[0m         normal_name,\n\u001B[1;32m    138\u001B[0m         input_tree\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[1;32m    139\u001B[0m         ancestral_reconstruction\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mFalse\u001B[39;00m,\n\u001B[1;32m    140\u001B[0m         chr_separator\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mX\u001B[39m\u001B[38;5;124m'\u001B[39m)\n\u001B[1;32m    142\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m (i\u001B[38;5;241m+\u001B[39m\u001B[38;5;241m1\u001B[39m) \u001B[38;5;241m%\u001B[39m (N_bootstrap\u001B[38;5;241m/\u001B[39m\u001B[38;5;241m/\u001B[39m\u001B[38;5;241m5\u001B[39m) \u001B[38;5;241m==\u001B[39m \u001B[38;5;241m0\u001B[39m:\n\u001B[1;32m    143\u001B[0m         logger\u001B[38;5;241m.\u001B[39minfo(\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m/\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m (\u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124m%\u001B[39m\u001B[38;5;124m) bootstrap runs completed\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;241m.\u001B[39mformat(i \u001B[38;5;241m+\u001B[39m \u001B[38;5;241m1\u001B[39m, N_bootstrap, \n\u001B[1;32m    144\u001B[0m                                                                   \u001B[38;5;28mint\u001B[39m((i\u001B[38;5;241m+\u001B[39m\u001B[38;5;241m1\u001B[39m)\u001B[38;5;241m/\u001B[39mN_bootstrap\u001B[38;5;241m*\u001B[39m\u001B[38;5;241m100\u001B[39m)))\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/core.py:59\u001B[0m, in \u001B[0;36mmain\u001B[0;34m(input_df, asymm_fst, maxcn, normal_name, input_tree, ancestral_reconstruction, chr_separator, prune_weight, allele_columns, wgd_x2, no_wgd, total_cn, n_cores, reconstruct_events)\u001B[0m\n\u001B[1;32m     54\u001B[0m     pairwise_distances \u001B[38;5;241m=\u001B[39m parallelization_calc_pairwise_distance_shrink_idea(sample_labels, asymm_fst, FSA_dict,\n\u001B[1;32m     55\u001B[0m                                                                             symbol_table, maxcn, chr_separator,\n\u001B[1;32m     56\u001B[0m                                                                             n_cores)\n\u001B[1;32m     57\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m     58\u001B[0m     \u001B[38;5;66;03m# pairwise_distances = calc_pairwise_distance_matrix(asymm_fst, FSA_dict)\u001B[39;00m\n\u001B[0;32m---> 59\u001B[0m     pairwise_distances \u001B[38;5;241m=\u001B[39m calc_pairwise_distance_shrink_idea(asymm_fst, FSA_dict, symbol_table, maxcn,\n\u001B[1;32m     60\u001B[0m                                                             chr_separator)\n\u001B[1;32m     62\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m (pairwise_distances \u001B[38;5;241m==\u001B[39m np\u001B[38;5;241m.\u001B[39minf)\u001B[38;5;241m.\u001B[39many()\u001B[38;5;241m.\u001B[39many():\n\u001B[1;32m     63\u001B[0m     affected_pairs \u001B[38;5;241m=\u001B[39m [(pairwise_distances\u001B[38;5;241m.\u001B[39mindex[s1], pairwise_distances\u001B[38;5;241m.\u001B[39mindex[s2])\n\u001B[1;32m     64\u001B[0m                       \u001B[38;5;28;01mfor\u001B[39;00m s1, s2 \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mzip\u001B[39m(\u001B[38;5;241m*\u001B[39mnp\u001B[38;5;241m.\u001B[39mwhere((pairwise_distances \u001B[38;5;241m==\u001B[39m np\u001B[38;5;241m.\u001B[39minf)))]\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/core.py:412\u001B[0m, in \u001B[0;36mcalc_pairwise_distance_shrink_idea\u001B[0;34m(model_fst, fsa_dict, symbol_table, max_chr_num, chrom_sep, parallel_run)\u001B[0m\n\u001B[1;32m    409\u001B[0m sample_a_fst \u001B[38;5;241m=\u001B[39m fsa_dict[sample_a]\n\u001B[1;32m    410\u001B[0m sample_b_fst \u001B[38;5;241m=\u001B[39m fsa_dict[sample_b]\n\u001B[0;32m--> 412\u001B[0m sample_a_shrink_fst, sample_b_shrink_fst \u001B[38;5;241m=\u001B[39m find_common_copy_number_profile_fsa(sample_a_fst, sample_b_fst, symbol_table, max_chr_num, chrom_sep)\n\u001B[1;32m    414\u001B[0m cur_dist \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mfloat\u001B[39m(fstlib\u001B[38;5;241m.\u001B[39mkernel_score(model_fst, sample_a_shrink_fst, sample_b_shrink_fst))\n\u001B[1;32m    415\u001B[0m pdm\u001B[38;5;241m.\u001B[39mloc[sample_a, sample_b] \u001B[38;5;241m=\u001B[39m cur_dist\n",
      "File \u001B[0;32m~/Project/Phylogenetic-tools/medicc2/medicc/core.py:328\u001B[0m, in \u001B[0;36mfind_common_copy_number_profile_fsa\u001B[0;34m(profile_1_fsa, profile_2_fsa, symbol_table, max_chr_num, chrom_sep)\u001B[0m\n\u001B[1;32m    325\u001B[0m profile_1_shrink \u001B[38;5;241m=\u001B[39m []\n\u001B[1;32m    326\u001B[0m profile_2_shrink \u001B[38;5;241m=\u001B[39m []\n\u001B[0;32m--> 328\u001B[0m max_chr_num \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mint\u001B[39m(max_chr_num)\n\u001B[1;32m    330\u001B[0m prev_label_1 \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m\n\u001B[1;32m    331\u001B[0m prev_label_2 \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m\n",
      "\u001B[0;31mTypeError\u001B[0m: int() argument must be a string, a bytes-like object or a real number, not 'Tree'"
     ]
    }
   ],
   "execution_count": 7
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": ""
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('medicc_env')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "0fe41a3dca095089b3d8143ec0d86970218fa4564a76fccbbd6486661a58c8d7"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
